ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm

```diff
@@ -0,0 +1,778 @@
+    # Instance parameters are described in a nlist
+    if ( exists($attr_value->{logFile}) ) {
+      $formatted_value .= " -l $attr_value->{logFile}";
+    }
+    if ( exists($attr_value->{configFile}) ) {
+      $formatted_value .= " -c $attr_value->{configFile}";
+    }
+    if ( exists($attr_value->{logKeep}) ) {
+      $formatted_value .= " -k $attr_value->{logKeep}";
+    }
+    
+  } elsif ( $value_fmt == LINE_VALUE_ARRAY ) {
+    $formatted_value = join " ", @$attr_value;
+
+  } elsif ( $value_fmt == LINE_VALUE_HASH_KEYS ) {
+    $formatted_value = join " ", keys(%$attr_value);
+    
+  } elsif ( ($value_fmt == LINE_VALUE_AS_IS) || ($value_fmt == LINE_VALUE_STRING_HASH) ) {
+    $formatted_value = $attr_value;
+    
+  } else {
+    $self->error("$function_name: invalid value format ($value_fmt) (internal error)")    
+  }
+
+  # Quote value if necessary
+  if ( ($line_fmt == LINE_FORMAT_PARAM) || ($line_fmt == LINE_FORMAT_ENVVAR) ) {
+    if ( $formatted_value =~ /\s+/ ) {
+      $self->debug(2,"$function_name: quoting value '$formatted_value'");
+      $formatted_value = '"' . $formatted_value . '"';
+    }
+  }
+  
+  $self->debug(2,"$function_name: formatted value >>>$formatted_value<<<");
+  return $formatted_value;
+}
+
+
+# This function formats a configuration line using keyword and value,
+# according to the line format requested. Values containing spaces are
+# quoted if the line format is not LINE_FORMAT_XRDCFG.
+#
+# Arguments :
+#        keyword : line keyword
+#        value : keyword value (can be empty)
+#        line_fmt : line format (see LINE_FORMAT_xxx constants)
+sub formatConfigLine () {
+  my $function_name = "formatConfigLine";
+  my $self = shift;
+
+  my $keyword = shift;
+  unless ( $keyword ) {
+    $self->error("$function_name: 'keyword' argument missing");
+    return 1;
+  }
+  my $value = shift;
+  unless ( defined($value) ) {
+    $self->error("$function_name: 'value' argument missing");
+    return 1;
+  }
+  my $line_fmt = shift;
+  unless ( defined($line_fmt) ) {
+    $self->error("$function_name: 'line_fmt' argument missing");
+    return 1;
+  }
+
+  my $config_line = "";
+
+  if ( $line_fmt == LINE_FORMAT_PARAM ) {
+    $config_line = "$keyword=$value";
+  } elsif ( $line_fmt == LINE_FORMAT_ENVVAR ) {
+    $config_line = "export $keyword=$value";
+  } elsif ( $line_fmt == LINE_FORMAT_XRDCFG_SETENV ) {
+    $config_line = "setenv $keyword = $value";
+  } elsif ( $line_fmt == LINE_FORMAT_XRDCFG_SET ) {
+    $config_line = "set $keyword = $value";
+  } elsif ( $line_fmt == LINE_FORMAT_XRDCFG ) {
+    $config_line = $keyword;
+    $config_line .= " $value" if $value;
+    # In trust (shift.conf) format, there should be only one blank between
+    # tokens and no trailing spaces.
+    $config_line =~ s/\s\s+/ /g;
+    $config_line =~ s/\s+$//;
+  } else {
+    $self->error("$function_name: invalid line format ($line_fmt). Internal inconsistency.");
```

Comment:
maybe use a readonly for the `internal error`/ `Internal inconsistency` type of messages. maybe we need a new reported `$self->deverror` for this type of messages (ie errors that have nothing to do with the system or service, but with the code), but these are different from configuration errors (e.g. when something is missing from the `$config`).
if you think this reasonable, open a issue in CAF for this.